### PR TITLE
Trigger user_login_failed event on unsucessful login.

### PR DESCRIPTION
**Pull request by kabalin**
This depends on #562.

### CODE REVIEW

File: classes/auth.php [hunk 1]
```diff
@@ -573,42 +573,56 @@ public function saml_login_complete($attributes) {
...
-        if (empty($attributes[$attr]) ) {
+        if (empty($attributes[$attr])) {
+            // Missing mapping IdP attribute. Login failed.
+            $event = \core\event\user_login_failed::create(['other' => ['username' => 'unknown',
+                'reason' => AUTH_LOGIN_NOUSER]]);
+            $event->trigger();
             $this->error_page(get_string('noattribute', 'auth_saml2', $attr));
         }
 
+        // Testing user's groups and allow access according to preferences.
```
Comment (danmarsden) [inline classes/auth.php:584]: Shouldn't this access check happen "After" we have worked out if we have a valid user?

Comment (kabalin) [reply classes/auth.php:584]: It does not need user record, it checks group in the attributes against plugin configuration.

Comment (kabalin) [reply classes/auth.php:584]: Possibly the comment can be improved.

Comment (danmarsden) [reply classes/auth.php:584]: of course - so in fact it makes a lot more sense to happen before we check if the user exists in Moodle - much better where you've put it there.

**Pull request merged**